File: tests/test_feature.py
import unittest
import random
from pysinthe import feature


class TestFeatureMethods(unittest.TestCase):

    def test_interactions(self):
        my_feature = feature.Feature("test",
                                     0,
                                     10,
                                     ["test", "test2"])
        self.assertTrue(my_feature.check_interaction("test2"))
        self.assertFalse(my_feature.check_interaction("test3"))


class TestPolymeraseMethods(unittest.TestCase):

    def setUp(self):
        self.pol = feature.Polymerase("mypol",
                                      40,
                                      30
                                      )
        self.pol2 = feature.Polymerase("mypol2",
                                       40,
                                       30
                                       )

    def test_init(self):
        self.assertEqual(self.pol.footprint, self.pol.stop - self.pol.start)
        self.assertEqual(self.pol.type, "polymerase")

    def test_move(self):
        old_start = self.pol.start
        old_stop = self.pol.stop
        self.pol.move()
        self.assertEqual(self.pol.start, old_start + 1)
        self.assertEqual(self.pol.stop, old_stop + 1)

    def test_move_back(self):
        old_start = self.pol.start
        old_stop = self.pol.stop
        self.pol.move_back()
        self.assertEqual(self.pol.start, old_start - 1)
        self.assertEqual(self.pol.stop, old_stop - 1)


class TestMaskMethods(unittest.TestCase):
    def setUp(self):
        self.mask = feature.Mask("mask",
                                 10,
                                 100,
                                 ["pol"]
                                 )

    def test_recede(self):
        old_start = self.mask.start
        self.mask.recede()
        self.assertEqual(self.mask.start, old_start + 1)


class TestElementMethods(unittest.TestCase):
    def setUp(self):
        self.element = feature.Element("myelement",
                                       23,
                                       60,
                                       ["ribosome"]
                                       )

    def test_save_state(self):
        self.element._covered = 3
        self.element.save_state()
        self.assertEqual(self.element._old_covered, 3)
        self.assertEqual(self.element._covered, self.element._old_covered)

    def test_cover(self):
        old_covered = self.element._covered
        self.element.cover()
        self.assertEqual(self.element._covered, old_covered + 1)

    def test_uncover(self):
        self.element._covered = 2
        self.element.uncover()
        self.assertEqual(self.element._covered, 1)
        # Make sure covering count doesn't go below zero
        self.element._covered = 0
        self.element.uncover()
        self.assertTrue(self.element._covered >= 0)

    def test_was_covered(self):
        self.element._covered = 0
        self.element.save_state()
        self.element.cover()
        self.assertTrue(self.element.was_covered())
        self.element.save_state()
        self.element.cover()
        self.assertFalse(self.element.was_covered())

    def test_was_uncovered(self):
        self.element._covered = 3
        self.element.save_state()
        self.element.uncover()
        self.element.uncover()
        self.element.save_state()
        self.assertFalse(self.element.was_uncovered())
        self.element.save_state()
        self.element.uncover()
        self.assertTrue(self.element.was_uncovered())

    def test_is_covered(self):
        self.element._covered = 1
        self.assertTrue(self.element.is_covered())
        self.element._covered = 0
        self.assertFalse(self.element.is_covered())


class TestTerminatorMethods(unittest.TestCase):
    def setUp(self):
        random.seed(22)
        self.term = feature.Terminator("myterm",
                                       23,
                                       60,
                                       {"rnapol": {"efficiency": 1.0},
                                        "ecolipol": {"efficiency": 0.6}
                                        }
                                       )
        self.pol = feature.Polymerase("rnapol",
                                      40,
                                      30
                                      )
        self.pol2 = feature.Polymerase("ecolipol",
                                       40,
                                       30
                                       )
        self.term.gene = "mygene"

        # Create temp termination signal
        self.pol.termination_signal.connect(lambda x: self.assertEqual(x, 60))

    def test_init(self):
        self.assertFalse(self.term.readthrough)


if __name__ == '__main__':
    unittest.main()

File: src/tests/test_feature.cpp
/* Copyright (c) 2017 Benjamin Jack All Rights Reserved. */

#include <catch.hpp>

#include <string>
#include <vector>

#include "feature.hpp"

TEST_CASE("Feature construction", "[Feature]")
{
  std::string name = "testing!";
  int start = 1;
  int stop = 10;
  std::vector<std::string> interactions = {"ecolipol", "rnapol"};
  Feature feat = Feature(name, start, stop, interactions);

  // Check that interactions have been set up correctly
  REQUIRE(feat.CheckInteraction("ecolipol") == true);
  REQUIRE(feat.CheckInteraction("some other pol") == false);
}

TEST_CASE("Polymerase construction and movement", "[Polymerase]")
{
  std::string name = "testing!";
  int footprint = 40;
  int speed = 60;
  Polymerase pol = Polymerase(name, footprint, speed);

  REQUIRE(pol.start() == 0);
  REQUIRE(pol.stop() == 39);

  pol.Move();
  REQUIRE(pol.start() == 1);
  REQUIRE(pol.stop() == 40);

  pol.MoveBack();
  REQUIRE(pol.start() == 0);
  REQUIRE(pol.stop() == 39);
}

TEST_CASE("Mask construction and movement", "[Mask]")
{
  std::string name = "testing!";
  int start = 1;
  int stop = 10;
  std::vector<std::string> interactions = {"ecolipol", "rnapol"};
  Mask mask = Mask(name, start, stop, interactions);

  // Check that interactions have been set up correctly
  REQUIRE(mask.CheckInteraction("ecolipol") == true);
  REQUIRE(mask.CheckInteraction("some other pol") == false);

  REQUIRE(mask.start() == 1);
  REQUIRE(mask.stop() == 10);

  mask.Recede();
  REQUIRE(mask.start() == 2);
  REQUIRE(mask.stop() == 10);
}

TEST_CASE("Element construction and state changes", "[Element]")
{
  std::string name = "testing!";
  int start = 1;
  int stop = 10;
  std::vector<std::string> interactions = {"ecolipol", "rnapol"};
  Element elem = Element(name, start, stop, interactions);

  SECTION("Saving the state")
  {
    elem.Cover();
    elem.Cover();
    elem.Cover();
    REQUIRE(elem.IsCovered());
    REQUIRE(elem.WasCovered());

    elem.SaveState();
    REQUIRE(!elem.WasCovered());

    elem.Uncover();
    elem.Uncover();
    elem.Uncover();
    REQUIRE(!elem.IsCovered());
    REQUIRE(elem.WasUncovered());

    elem.SaveState();
    REQUIRE(!elem.WasUncovered());
  }

  SECTION("Uncovering and covering an element")
  {
    // Make sure covering counts don't go below zero
    elem.Uncover();
    elem.Uncover();
    elem.Uncover();
    REQUIRE(!elem.IsCovered());
    REQUIRE(!elem.IsCovered());

    elem.Cover();
    REQUIRE(elem.IsCovered());
    REQUIRE(elem.WasCovered());

    elem.Uncover();
    REQUIRE(!elem.IsCovered());
    REQUIRE(!elem.WasCovered());
    REQUIRE(!elem.WasUncovered());
  }
}

TEST_CASE("Terminator construction", "[Terminator]")
{
  std::string name = "testing!";
  int start = 1;
  int stop = 10;
  std::vector<std::string> interactions = {"ecolipol", "rnapol"};
  Terminator term = Terminator(name, start, stop, interactions);
  REQUIRE(!term.readthrough());
  REQUIRE(term.reading_frame() == -1);
  // Test for reading frame checking
  term.set_reading_frame(1);
  REQUIRE(term.CheckInteraction("ecolipol", 1));
  REQUIRE(!term.CheckInteraction("ecolipol", 0));
  REQUIRE(!term.CheckInteraction("nopol", 1));
}
